src/sentry/tasks/releasemonitor.py

```diff
@@ -0,0 +1,178 @@
+                        ],
+                        where=[
+                            Condition(
+                                Column("started"), Op.GTE, datetime.utcnow() - timedelta(hours=6)
+                            ),
+                            Condition(Column("started"), Op.LT, datetime.utcnow()),
+                            Condition(Column("org_id"), Op.EQ, org_id),
+                            Condition(Column("project_id"), Op.IN, project_ids),
+                        ],
+                        granularity=Granularity(21600),
+                        orderby=[OrderBy(Column("org_id"), Direction.ASC)],
+                    )
+                    .set_limit(CHUNK_SIZE + 1)
+                    .set_offset(offset)
+                    .set_totals(True)
+                )
+                results = snuba.raw_snql_query(
+                    query, referrer="tasks.process_projects_with_sessions"
+                )
+                data = results["data"]
+                count = len(data)
+                more_results = count >= CHUNK_SIZE
+                offset += count
+                total_sessions = results["totals"]["sessions"]
+
+            # 3. Using the sums from #2, calculate adoption rate (relevant sessions / all sessions) update the appropriate ReleaseProjectEnvironment model adopted/unadopted fields.
+            with metrics.timer(
+                "sentry.tasks.monitor_release_adoption.process_projects_with_sessions.updates"
+            ):
+                for row in data:
+                    try:
+                        if row["sessions"] / total_sessions >= REQUIRED_ADOPTION_PERCENT:
+                            rpe = ReleaseProjectEnvironment.objects.get(
+                                project_id=row["project_id"],
+                                release_id=Subquery(
+                                    Release.objects.filter(
+                                        organization=org_id, version=row["release"]
+                                    ).values("id")[:1],
+                                ),
+                                environment__name=row["environment"],
+                                environment__organization_id=org_id,
+                            )
+                            adopted_ids.append(rpe.id)
+                            if rpe.adopted is None:
+                                rpe.update(adopted=timezone.now())
+                    except ReleaseProjectEnvironment.DoesNotExist:
+                        logger.info(
+                            "process_projects_with_sessions.skipped_update",
+                            extra={
+                                "project_id": row["project_id"],
+                                "environment": row["environment"],
+                                "release": row["release"],
+                                "reason": "ReleaseProjectEnvironment.DoesNotExist",
+                            },
+                        )
```

Comment:
The ghost of @JTCunning would tell us to put this in our product rather than logs. Also might be worth tracking this as a metric so we can track how often it happens

Reply:
I would do a lot to honour the ghost of @JTCunning